x/evm/ante/fee.go

```diff
@@ -0,0 +1,110 @@
+package ante
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	accountkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	paramskeeper "github.com/cosmos/cosmos-sdk/x/params/keeper"
+	"github.com/ethereum/go-ethereum/consensus/misc/eip4844"
+	evmkeeper "github.com/sei-protocol/sei-chain/x/evm/keeper"
+	evmtypes "github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+type EVMFeeCheckDecorator struct {
+	evmKeeper     *evmkeeper.Keeper
+	paramsKeeper  *paramskeeper.Keeper
+	bankKeeper    bankkeeper.Keeper
+	accountKeeper *accountkeeper.AccountKeeper
+}
+
+func NewEVMFeeCheckDecorator(evmKeeper *evmkeeper.Keeper, paramsKeeper *paramskeeper.Keeper) *EVMFeeCheckDecorator {
+	return &EVMFeeCheckDecorator{
+		evmKeeper:     evmKeeper,
+		paramsKeeper:  paramsKeeper,
+		bankKeeper:    evmKeeper.BankKeeper(),
+		accountKeeper: evmKeeper.AccountKeeper(),
+	}
+}
+
+func (fc EVMFeeCheckDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// only check fee in CheckTx (similar to normal Sei tx)
+	if !ctx.IsCheckTx() || simulate {
+		return ctx, nil
+	}
+
+	txData, found := evmtypes.GetContextTxData(ctx)
+	if !found {
+		return ctx, errors.New("could not find eth tx when checking fees")
+	}
+
+	ver, ok := evmtypes.GetContextEVMVersion(ctx)
+	if !ok {
+		return ctx, errors.New("could not find eth version when checking fees")
+	}
+
+	// if EVM version is London or later, gas fee cap will be used to cap the overall fee consumption,
+	// so we need to make sure that cap is at least as large as the required base fee
+	if ver >= evmtypes.London && txData.GetGasFeeCap().Cmp(fc.getBaseFee(ctx, txData.GetGas())) < 0 {
```

Comment:
yeah that's true. I'll just include London and Cancun here (since Cancun hasn't landed on ethereum mainnet yet we should probably start with the current version)